### PR TITLE
Add support for static boxes in 'leveled' type; add node box types 'leveled_plantlike[_rooted]'

### LINKED ISSUE #8202: Selectionbox of plantlike_rooted is constant while plant height might vary
##### Issue type
- Design flaw? / bug?
 
##### Minetest version
5.0.0-dev

##### Summary
The height of `plantlike_rooted` depends on `param2`, but the selection box size is constant and ignores the `param2` completely. But the height of the plant can in theory vary from 1 to 16. In the worst case, modders are forced to use a single (!) selection box for a node that might have a plant of 16 different heights.

This doesn't seem right.

##### Suggestion
I think the selection box height should be dynamic and directly related to `param2`. Or at least it should be possible to opt out of fixed selection box height.

### PR DESCRIPTION
**Pull request by Wuzzy2**
## The problems

Currently, 'leveled' boxes will allways have their upper face change with param2. So this only allows you to do fairly simple things. If you need a fully static box as part of a leveled nodebox, you are out of luck.

Also, you can't get selection boxes to be in sync with leveled plantlike nodes. No matter what you do, they are always off.

## The PR

This PR adds a new field '`leveled_fixed`' for the nodebox type 'leveled'. This is an optional box (or boxes) for leveled-type nodeboxes, but it will be fully static.

This screenshot shows an example (it's always the same node, but with different param2): https://satoshiupload.com/images/q5VZ1MoO8I.png

Moreover, it adds two new nodebox types `leveled_plantlike` and `leveled_plantlike_rooted`. These work like `leveled` except they are in sync with the height of the leveled 'plant' for the drawtypes `plantlike` and `plantlike_rooted`, respectively. This allows you to make pixel-perfect selection boxes for such nodes.

The height, however, is capped at one point because leveled plantlike nodes can get very high and I'm not sure how reliable selection boxes are when they're really high.

Fixes #8202.

## Use cases
* Cauldron with water of variable height (`leveled_fixed` is the cauldron (which is static), `fixed` is the water box (which varies in height))
* Trash can with something inside
* Some kind of simple ice spike
* Kelp plant with varying height with pixel-perfect selection box that respects height

## Please note

Unfortunately, the naming is a little bit weird for `leveled` nodeboxes. For some reason, the `fixed` box is actually NOT fixed, since it varies in height. The new nodebox field `leveled_fixed` is the box (or boxes) that is ACTUALLY fixed.

It has been like that before, I cannot change that trivially w/o breaking compability. I think it is best to just keep the naming for the sake of compability, which is MUCH more important.

The nodebox types `leveled_plantlike` and `leveled_plantlike_rooted` are intended to be used as selection boxes so you can capture the plant pixel-perfectly. New nodebox types were neccessary because `leveled` has a step of 1/64 while leveled plantlike nodes have a step of 1/16. The new nodebox types use a step of 1/16 so it matches. `leveled_plantlike_rooted` is basically the same as `leveled_plantlike` but a height offset is applied (because of the base cube).
Also, unlike `leveled`, the new nodebox types do NOT restart at param2=128.

For collision boxes, they are only safe to use if you keep the upper end of the "plant" below a height of -0.5+127/64 (roughly 1.5), otherwise, collision bugs will start to occur. There is nothing in the code that enforces this limit so it's the responsibilty of the game author to make sure the "plant" doesn't become too tall. However, using these nodebox types as collision boxes is a rather esoteric use case anyway, so it should not be that big of a deal. I documented that anyway, just in case.

## How to test

Use the following test nodes in DevTest. Try them together with the Param2 Tool.

* `testnodes:nodebox_leveled_fixed` (new node)
* `testnodes:plantlike_leveled` (old node, new dynamic selection box)
* `testnodes:plantlike_rooted_leveled` (old node, new dynamic selection box)

### CODE REVIEW

**Comment by Wuzzy2** [comment]
I've got a question: Would it be OK to allow selection boxes to be arbitrarily high? A leveled plantlike node (`testnodes:plantlike_leveled` in DevTest) can reach a maximum Y of 15.4375 (at `param2=255`). I have made a few tests with dropping the height limits I enforced (currently at max. Y of 3.5) and the selection box still worked flawlessly even at max. height. I can punch the node at any height just fine and if I place something at the selection box, even at max height, it works like you expect (i.e. the node is placed at the correct height, not just at the bottom).

So, can I drop the height limit safely and allow for selection boxes to go for the full plantlike height? Or is there still some edge case I might have overlooked?

Dropping the hight limit would be really neat because this means you could make e.g. algae, kelps, etc. pointable everywhere, and that pixel-perfectly.

**Comment by loffren174** [comment]
I like all of these ideas to extend the functionality of param, param2, param3,
especially this one, and those allowing more simultaneous rotations+color bits.
This one in particular would be especially neat for any sort of 3d-pillar-progress bar/energy bar/fluid fill status node, like making sci-fi pipes that represent a counter or energy threshold, or simply showing a player their score in a minigame with a neat pillar. 
Sort of like this stock image, but rectangular:
https://images.stockfreeimages.com/592/sfixl/5925868.jpg
With this pr it would be trivial to make any voluminous fluid-holding tank represent a linear value by its color height, without sacrificing the ability to texture its borders.

**Comment by rubenwardy** [comment]
Unfortunately, this feature is not on the roadmap and hasn't received a concept approval. Under the [new rules](https://github.com/minetest/minetest/blob/master/.github/CONTRIBUTING.md#code), non-roadmap non-bugfix PRs have 7 days to be concept approved, otherwise they should be closed. These PRs can also have preapproval if they fix an issue that is concept approved

If a core dev wants to support this PR, they may reopen it and apply the "Supported by core dev" label

**Comment by Wuzzy2** [comment]
I totally understand that PRs can take months or even years to be even recognized by the core devs. I understand that time is a limited resource. I also understand when bad PRs are rejected for being bad. Because eventually, with enough patience, most of my PRs came through eventually. And those that were ultimately rejected were always rejected for a reason. Even if I didn't always agree with it, at least there was *a* reason. The long waiting times weren't great, but at least patience pays off eventually.

But this new policy is just about murdering PRs without a single core dev having even *looked at it*. This is infuriating! There's no review, no consideration, no explanation, not even a single topic about the mere concept. Just a giant "Fuck you!", as if my work is not even taken serious anymore. Do you realize how condescending this looks? And no, I don't buy the excuse of "If a core dev wants to support this PR" ... It is closed now. Who is going to regularily look at the list of *closed* PRs?

And a week seems short. So if all or most core devs are on vacation or distracted, all PRs posted in this period are by definition already doomed. Frankly, this new policy sucks.

I am kinda pissed … But OK, I play along for now … According to the roadmap, I am supposed to post an issue before a PR. So I am going to do that … Among the >900 issues. Ooff. Fun times! I predict this will slow down development even further, because there are now two layers of approval (one for the issue, one for the PR), as if Minetest doesn't already have enough bureaucracy. ;-)

Also: How on earth is "improving drawtypes" *not* on the roadmap? I am utterly confused.

**Comment by v-rob** [comment]
Well, I myself feel unhappy to let this PR get closed because it seems like a really good idea, but if I add a "Supported by core dev" label, then it seems as though I'm supposed to take a high level of responsibility for it.  It seems to me that it's more likely for PRs to get closed than for core devs to take more on to their plate.

But whatever.  I'm not letting this get closed; I think Minetest's most important asset is nodes, so PRs to improve them are very good.  I guess that basically makes me one forced reviewer.

**Comment by Wuzzy2** [comment]
I have just rebased this PR to make it testable again. I repeat my question from before:

> Would it be OK to allow selection boxes to be arbitrarily high?

See <https://github.com/minetest/minetest/pull/11391#issuecomment-868978981>

**Comment by v-rob** [comment]
I'd personally remove the limit.  There's no limit to selection boxes (or collision boxes) for any other node, so it doesn't make much sense to enforce it for this type in specific.  Of course, collision boxes break after a certain size, and even nodeboxes don't always display right at huge sizes, so if huge selection boxes break, it wouldn't be much of a surprise (or much of a problem either).

**Comment by Wuzzy2** [comment]
Thanks for that. I am waiting for more opinions on this.

In my tests so far, very large selectionboxes (if I disable the limit) still work fine. I'm mostly askin in case I overlook something.
Very large collisionboxes break (of course, that's not new and by design), but this is documented.

**Comment by kab0u** [comment]
I'd like to express my support for this feature, but with some caveats.  

I am currently creating a ["composter"](https://git.minetest.land/kabou/MineClone2/src/branch/composters/mods/ITEMS/mcl_composters) node for MineClone2 and its derivatives and some aspects of this node box type might fit my purpose very well.  However, it would not entirely fit my purpose, because I also need to have different textures depending on the node's level.  AFAICS there is no support for having different textures mapped to varying levels in this PR.  

The composter is a static box that can be filled with 7 different levels of compost. It uses three different top textures for the node, ie. a "bottom" texture for the empty composter, a "compost" texture for the various stages of filling, and a "compost with bone meal" texture when the composter is ready for harvest.  The solution that I am using now is to have several different node boxes defined each with a node attribute "_compost_level" and reading that from minetest.registered_nodes[node.name][_compost_level] .  Seven nodes have almost identical definitions, apart from the node box's height and the name and _compost_level attribute that have been parametrized in the node registration.  The "empty" and "ready" composters have additionally different textures defined.

If I were to use the static leveled option that this PR implements, I could set the level of individual instances of composter nodes, but if I were to update the texture in the node definitions, this would affect the textures of *all* composter nodes placed on the map.  

Hence my additional feature request for some form of level-indexed node textures.  Possibly, the original leveled type could even be extended to have indexed textures, that would IMHO add even more value to this PR.

**Comment by Wuzzy2** [comment]
Please don't hijack PRs with major feature requests.

**Comment by v-rob** [comment]
Textures per nodebox (or per level) is something I've thought of myself before as there's no other way of doing that sort of thing without a model.  However, it's not something for this PR, without a doubt.

(Now I really need to review this PR...)

**Comment by kab0u** [comment]
Apologize if my remarks were taken as an intent to hijack the PR.  I just wanted to describe a use case beyond the trivial and suggest a possible extension/enhancement.  If that is beyond the scope of this PR then so be it.

**Comment by v-rob** [comment]
I did some extreme selection box testing.  I tested selection boxes with heights up to 220 nodes.  If anything would break selection boxes, the camera offset would.  At a height of 220, it will 100% cross mapblock boundaries and camera offset boundaries.  Surprisingly, selection boxes still seem to work perfectly.  The bad news: selection boxes that big obliterated my framerate.  I was getting perhaps 0.5-1 FPS.  It doesn't matter whether the node exists on the map or not: just having a node definition with such a selection box causes it.

Now, the highest selection box leveled plantlike_rooted could achieve (in this PR without the limit) is a height of just below 8.  I didn't take any precise measurements, but there seemed to be very little difference in terms of framerate between Minetest with a selection box that big and Minetest without.  In fact, I couldn't tell any difference up until after 14 or so.  It might be that performance doesn't drop until selection boxes become bigger than mapblock sizes.  In any case, I feel like removing the limit would be reasonable.

**Comment by Wuzzy2** [comment]
Thanks for testing. So I have increased the limit to 16.5, which is slightly above the maximum possible height of a `plantlike_rooted` node (measured from the center of the node pos).

It works fine for me, no issues or FPS drop. It also doesn't matter if the "plant" crosses mapblock boundaries, even if `param2=255`.

If you require any other changes or bugfixes, let me know!

**Comment by Zughy** [comment]
@Wuzzy2 please rebase

**Comment by Wuzzy2** [comment]
Rebase done. However, I did not test intensively. The rebase was a bit more tricky this time, so there might be mistakes.

File: src/mapnode.cpp
Comment (Desour) [inline src/mapnode.cpp:407]: This code could be much shorter btw. if you rotate by facedir first and then by axisdir.
See: 
https://github.com/minetest/minetest/blob/0530ec11c0371bbe10a22c1800e65ad10234c0b8/src/client/mesh.cpp#L298-L314

**Comment by Wuzzy2** [comment]
@Desour: That code is mostly inherited from the previous code, just sayin’.

File: src/mapnode.cpp
Comment (v-rob) [inline src/mapnode.cpp:282]: Some wacky indentation with spaces here.

**Review by v-rob — COMMENTED** [review]
Tested and works.  Code looks pretty much fine; the diff is actually quite simple if you hide whitespace changes.  I have only one real question about functionality, and if that gets answered/resolved, approval from me.

**Comment by Wuzzy2** [comment]
@v-rob: Done.

**Review by v-rob — APPROVED** [review]

**Comment by Zughy** [comment]
@Wuzzy2 please rebase

**Comment by Wuzzy2** [comment]
Rebase done. The rebase was difficult again, I hope I didn't screw up. Surface-level testing looks OK.

Let's hope this PR gets merged before hitting the 2 year anniversary. XD

**Comment by Wuzzy2** [comment]
Done.

**Comment by Wuzzy2** [comment]
The rebase has been completed. I also noticed the documentation was slightly wrong/outdated, so I fixed it.

File: src/nodedef.cpp
Comment (appgurueu) [inline src/nodedef.cpp:1262]: May I suggest: `f32 max = std::max({fabsf(half_processed.MinEdge.X), ...})` (`std::max` with an initializer list). That should allow you to get rid of the loop.

File: src/nodedef.cpp
Comment (appgurueu) [inline src/nodedef.cpp:1259]: Shouldn't this be an `else if`?

File: src/nodedef.cpp
Comment (appgurueu) [inline src/nodedef.cpp:1249]: This seems slightly suspicious to me: What if you have a min point larger than 0, or a max point less than 0? I would initialize this box to `{INFINITY, INFINITY, INFINITY, -INFINITY, -INFINITY, -INFINITY}` instead.

**Review by appgurueu — COMMENTED** [review]
Thanks for the PR.

Tested the three devtest test nodes, they work as expected - boxes match the visuals and keep matching when changing param2. Also tested with the box visualizer tool that the collision box is correct for the leveled fixed test node. The fixed boxes of e.g. stairs still work as expected.

I have some comments on the implementation, but nothing major.

**Comment by Wuzzy2** [comment]
Sorry for the long wait, but I finally worked on addressing @appgurueu's comments. I changed the code and added some documentation clarification.

There is no rebase yet (I'll wait for further feedback first).

Here are my responses to the comments:

> Also, is the safe selection box limit really that large (~16 nodes)? This would seem unlikely to me, wouldn't raycasts have to be much slower if this was the case?

Oh crap. I haven't even thought of raycasts. This might ruin the entire PR because allowing high selection boxes is the main "selling point" of this PR. That feature is required to do what I want.

I think the performance needs to be tested to see how large the impact is.

(I wish we had raycasts that work on collisionboxes as well.)

> Why? If this is to silence a compiler warning, 

Pretty much that. It's standard practice to always have the default case, even if it does nothing.

> Why not switch on nodebox.type here, instead of an if-else if-else chain?

I think a switch would look messier in this instance.

> Shouldn't this be an else if?

No. This branch is about paramtype2, the previous conditional is about the nobebox type.

> This seems slightly suspicious to me: What if you have a min point larger than 0, or a max point less than 0? I would initialize this box to {INFINITY, INFINITY, INFINITY, -INFINITY, -INFINITY, -INFINITY} instead.

Your suggestion breaks the feature. But yeah, this might probably need more testing/investigation to be sure everything is OK.

**Comment by appgurueu** [comment]
> I'll wait for further feedback first

Good idea, let's talk a bit about this first.

> Oh crap. I haven't even thought of raycasts. This might ruin the entire PR because allowing high selection boxes is the main "selling point" of this PR. That feature is required to do what I want.

Looking at the implementation, it should work correctly, since we use the bounds (i.e. how large a node can be) to search the area where relevant nodes may be.

> I think the performance needs to be tested to see how large the impact is.

Node size is a linear factor. If node boxes were at most 4 nodes tall before and now they can be 16 nodes tall, raycast performance - with respect to nodes - can be expected to decrease by roughly a factor of 4. If it was 2 nodes before (MTG doors) I would expect it to become 8 times slower *if this feature is used*. We should measure this however.

I haven't tested this yet, but it should be testable with [xcam](https://content.minetest.net/packages/erlehmann/xcam/).

I don't think large selection or collision boxes for nodes are a particularly good idea. Even if used sparsely, our current architecture doesn't really allow dealing with them efficiently (or in a simple way, from a mod perspective). I would try to not stray too far from "1 node = 1 metric unit, cubed, or maybe a little less, but ideally no more". (With that in mind, have you considered a variant of this where large plants would consist of many nodes?)

(Even smaller nodes like doors already create problems, since they break simplifying assumptions mods - and the engine - may make. For example: Can I now place a node overlapping with such a large plant? For water this may be desired, but what if I place, say, dirt?)

Related: https://github.com/minetest/minetest/issues/14480

> (I wish we had raycasts that work on collisionboxes as well.)

If you're thinking of the same use cases I'm thinking of, you would probably prefer something that lets you check for collisions with volumes, such as node boxes (to be placed), no? That's what I implemented in [dbwts](https://github.com/appgurueu/disable_build_where_they_stand) for example. Nowadays this could be implemented in much less code by leveraging grorp's new collision box getter API. (But I agree that we should probably have utilities akin to raycasts for this in the engine, for example to let modders check whether there is space to spawn an entity somewhere.)

---

> Pretty much that. It's standard practice to always have the default case, even if it does nothing.

Not very important, but I don't think that it is in any way a "good" standard practice to have a no-op default case.

If the `default` case shouldn't happen (e.g. is a violation of invariants), I'd put something like `assert(false);` or `throw std::runtime_error("invalid enum value");` there.

I try to keep in mind what happens if the code is changed (e.g. an enum value is added): Ideally, that should produce a compiler warning everywhere where it might matter. Explicitly specifying a `default` case prevents this.

If a compiler warning is not possible, it should fail loudly at runtime, rather than silencing the issue and doing something which may or may not be correct.

> Your suggestion breaks the feature. But yeah, this might probably need more testing/investigation to be sure everything is OK.

I took a closer look and this is used only by raycasts, to get an upper bound on selection boxes. I see no issue with always including the origin in that case. So we can scratch that point.

**Comment by Wuzzy2** [comment]
Here's some context for this PR: The main reason why I added this PR is tall underwater plants like kelp and algae. Nodes that are `plantlike_rooted` and `leveled`. It always annoyed me the selection box rarely matches the actual plant but we had to resort to a fixed-size selectionbox. The selection box is usually too small or too large. If you want to dig a large kelp, you have to go all the way down.

But I understand the performance concerns. Leveled plants can get *really* tall. I'm looking forward to the performance test.

If it turns out performance really goes to shit, I have an idea to "rescue" this PR:

By limiting the selection box height to our current maximum. So smaller kelps still have a pixel-perfect selection box, but very tall kelps have their selection box capped at whatever is our current max height. This unfortunately breaks my vision for this PR somewhat, but it seems the best compromise to not break performance. Note that this PR only affects the Y size. X and Z are unaffected.

> If the default case shouldn't happen (e.g. is a violation of invariants), I'd put something like assert(false); or throw std::runtime_error("invalid enum value"); there.

Ah yes, this is a good idea.

**Comment by sfence** [comment]
Builded, tested.
Selection boxes do not follow the leveled plantlike nodes in my case (linux).
Nodes `testnodes:plantlike_leveled` and `testnodes:plantlike_rooted_leveled` were placed and the "Param2 Tool" was applied to them.
screenshot_20240805_181234
screenshot_20240805_181353

**Comment by sfence** [comment]
Sorry for the mistake. Works as expected.
Rebased to actual master -> easier building process.

**Comment by Zughy** [comment]
@sfence was that an approval?

**Comment by sfence** [comment]
> @sfence was that an approval?

No, I do not want to approve this, because there are still comments from @appgurueu which are not resolved.

**Comment by appgurueu** [comment]
> I'm looking forward to the performance test.

Tested with MTG.

As expected the impact is linear. Before, the selection box int union was `-1 -1 -1 1 3 1` (due to kelp). Now it is `-1 -1 -1 1 16 1`. So according to the theory runtime could roughly quadruple. 

I tested this practically and it's not that bad because other overheads which are independent of this search area size contribute a significant factor. Still in my testing this **doubles raycast runtime**: What took 0.25s before now takes 0.5s. Registering any node that uses this selection box suffices for this. Whether the node is used at all, or only used with smaller param2s, is irrelevant.

Ultimately this is not my only concern though. I feel like this implementation of kelp and algae as a single node which is awfully tall really is an ugly hack; it simply breaks the node grid too much. This causes issues in all areas - selection, collision, rendering, mod logic. Nodes like doors already cause enough issues in mods, and this is much worse. Frankly if we had a decent alternative ready I'd love to deprecate these 16-node tall nodes.

<details>
<summary>ceci ne sont pas seize bloques</summary>

Screenshot from 2024-08-28 14-12-38

Screenshot from 2024-08-28 13-59-22
</details>

That said I do see the point that (1) since we have this hack already and it has its valid use cases, we might as well fix it; (2) the performance tradeoff is ultimately up to modders / game devs.

This PR works. But modders should perhaps be warned that this isn't good for the performance of everything that respects these boxes, and likely never will be.

**Comment by appgurueu** [comment]
> Did you also test the case when the game doesn't use such leveled nodes?

That's exactly what I tested to arrive at this conclusion:

> Still in my testing this doubles raycast runtime: What took 0.25s before now takes 0.5s.

**Comment by sfence** [comment]
I cannot reproduce the raycast performance problem on my Arch Linux builds -> this PR vs the actual master.

The only problem is the first few runs are typically quicker for big areas. Probably caused by processing unloaded blocks at beginning.

**Comment by appgurueu** [comment]
> I cannot reproduce the raycast performance problem on my Arch Linux builds -> this PR vs the actual master.

How specifically are you benchmarking? Are you making sure that a node using this feature is registered? What is the base game you're testing against?

I doubt that this has no performance implications because the logic strongly suggests otherwise. With this PR something like 4x the nodes have to be checked.

I tested using the following code on MTG. Essentially it just registers a node using this feature, which is necessary for the selection box union to grow. I then used the very rudimentary `/rays` chatcommand below to get some rough numbers. You can repeat this multiple times to check that the numbers are consistent. For minimally better reproducability you might want to seed the random. Also make sure not to move of course.

<details>

```lua
minetest.register_node("dev:kelp", {
	description = "Kelp or something",
	drawtype = "nodebox",
	paramtype = "light",
	paramtype2 = "leveled",
	node_box = {
		type = "leveled_plantlike",
		fixed = { -0.4, -0.5, -0.4, 0.4, 0.4, 0.4 },
	},
	collision_box = {
		type = "leveled_plantlike",
		fixed = { -0.4, -0.5, -0.4, 0.4, 0.4, 0.4 },
	},
	groups = {dig_immediate=3},
	place_param2 = 255,
})

minetest.register_chatcommand("rays", {
	description = "Do the rays",
	func = function(name)
		local t = minetest.get_us_time()
		local p = minetest.get_player_by_name(name)
		for i = 1, 1000 do
			for thing in minetest.raycast(p:get_pos(), p:get_pos():offset(math.random(-100, 100), math.random(-100, 100), math.random(-100, 100)), false, false) do
			end
		end
		return true, (minetest.get_us_time() - t) .. "us" 
	end
})
```

</details>

**Comment by sfence** [comment]
@appgurueu 
With this scenery, I see it too.
Interestingly, it takes visible no effect for one run.

**Comment by Wuzzy2** [comment]
I have decided to revive this PR under #15959.

It only includes `leveled_fixed`, the plantlike nodebox types have been removed. Hopefully this will make is more likely to accept this feature.

**Pull request closed without merging**
